src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,212 @@
     public void add(T value, int index) {
+        checkAddIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else if (index == 0) {
+            linkFirst(value);
+        } else {
+            link(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = getNodeByIndex(index).value;
+        getNodeByIndex(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(index);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int removeByIndex = getIndexIfContains(object);
+        if (removeByIndex == -1) {
+            return false;
+        }
+        unlink(removeByIndex);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkLast(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
+    }
+
+    private void linkFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        newNode.next = head;
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void link(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        Node<T> previous = getNodeByIndex(index - 1);
+        newNode.next = previous.next;
+        previous.next.prev = newNode;
+        previous.next = newNode;
+        newNode.prev = previous;
+        size++;
+    }
+
+    private int getIndexIfContains(T object) {
+        int index = -1;
+        Node<T> current = head;
+        int iteration = 0;
+        while (current != null) {
+            if (current.value == object || current.value != null && current.value.equals(object)) {
+                return iteration;
+            }
+            current = current.next;
+            iteration++;
+        }
+        return index;
+    }
+
+    private T unlink(int index) {
+        if (size == 1) {
+            return unlinkTheOne();
+        } else if (index == 0) {
+            return unlinkFirst();
+        } else if (index == size - 1) {
+            return unlinkLast();
+        } else {
+            return unlinkWithin(index);
+        }
+    }
```

Comment:
Can you simplify this method instead of having a separate logic for each  case?

Reply:
I'm not sure I've got you right) I simplified 'if-else' construction by removing the redundant 'else-if' condition.